sdk/typespec/typespec_client_core/src/http/headers/mod.rs

```diff
@@ -9,42 +9,66 @@ mod microsoft;
 pub use common::*;
 pub use microsoft::*;
 
-use std::{borrow::Cow, fmt::Debug, str::FromStr};
+use std::{borrow::Cow, convert::Infallible, fmt::Debug, str::FromStr};
 use typespec::error::{Error, ErrorKind, ResultExt};
 
 /// A trait for converting a type into request headers.
 pub trait AsHeaders {
+    type Error;
     type Iter: Iterator<Item = (HeaderName, HeaderValue)>;
-    fn as_headers(&self) -> Self::Iter;
+
+    fn as_headers(&self) -> Result<Self::Iter, Self::Error>;
 }
 
 impl<T> AsHeaders for T
 where
     T: Header,
 {
+    type Error = Infallible;
     type Iter = std::vec::IntoIter<(HeaderName, HeaderValue)>;
 
     /// Iterate over all the header name/value pairs.
-    fn as_headers(&self) -> Self::Iter {
-        vec![(self.name(), self.value())].into_iter()
+    fn as_headers(&self) -> Result<Self::Iter, Self::Error> {
+        Ok(vec![(self.name(), self.value())].into_iter())
     }
 }
 
 impl<T> AsHeaders for Option<T>
 where
     T: AsHeaders<Iter = std::vec::IntoIter<(HeaderName, HeaderValue)>>,
 {
+    type Error = T::Error;
     type Iter = T::Iter;
 
     /// Iterate over all the header name/value pairs.
-    fn as_headers(&self) -> Self::Iter {
+    fn as_headers(&self) -> Result<Self::Iter, T::Error> {
         match self {
             Some(h) => h.as_headers(),
-            None => vec![].into_iter(),
+            None => Ok(vec![].into_iter()),
         }
     }
 }
 
+/// Extract a value from the [`Headers`] collection.
+///
+/// The [`FromHeaders::from_headers()`] method is usually used implicitly, through [`Headers::get()`] or [`Headers::get_optional()`].
+pub trait FromHeaders: Sized {
+    type Error: std::error::Error + Send + Sync + 'static;
+
+    /// Holds a list of the header names that [`FromHeaders::from_headers`] expects.
+    ///
+    /// Used by [`Headers::get()`] to generate an error if the headers are not present.
+    const HEADER_NAMES: &'static [&'static str] = &[];
+
+    /// Extracts the value from the provided [`Headers`] collection.
+    ///
+    /// This method returns one of the following three values:
+    /// * `Ok(Some(...))` if the relevant headers are present and could be parsed into the value.
+    /// * `Ok(None)` if the relevant headers are not present, so no attempt to parse them can be made.
+    /// * `Err(...)` if an error occurred when trying to parse the headers. This likely indicates that the headers are present but invalid.
+    fn from_headers(headers: &Headers) -> Result<Option<Self>, Self::Error>;
```

Comment:
The age-old question of how to represent something that is optional **and** fallible. I'm open to other Options that get us to the same Result (puns absolutely intended).

If we're going to have nested Result/Option, I think this is the best ordering, because it allows you to write `if let Some(h) = from_headers(...)?`. I think this is the most ergonomic option.

We do have `Headers::get()` which coerces errors in to `azure_core::Error` and produces an `ErrorKind::DataConversion` error if the required headers are missing, which can be used for reading mandatory header.